### PR TITLE
Add Authorization using Pundit

### PR DESCRIPTION
**Pull request by srb-**
- Adds Auth policy via the `pundit` gem
- Creates a sample policy for `aircraft` creation
- Permissions stored in a `context` field on the `User` object
- Modified the Aircraft `show` view to have policy-dependent behavior
- Modified the User views to have a virtual attribute `wing_admin` so we can easily toggle users in and out of admin to check permissions
  - used a constant for the permission - obviously if we did real RBAC we'd make some sort of dynamic list

**Testing**: 
- manual, and added unit tests

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -12,5 +12,5 @@ The purpose of this project is to determine what changes will be need to made on
 7. Run `bin/rails s -b 0.0.0.0` and go to the URL specified in your terminal
 
 ### To have some Data to Play
-8. Run `rails db:seed`
+1. Run `rails db:seed`
```
Comment (srb-) [inline README.md:15]: To stop a Rubocop complaint.

File: config/docker/docker-compose.dev.yml [hunk 2]
```diff
@@ -43,6 +43,7 @@ services:
       - /tmp
     networks:
       - noticed-flight-sim-net
+    command: bash -c "bin/rails s -p 3000 -b '0.0.0.0'"
```
Comment (srb-) [inline config/docker/docker-compose.dev.yml:46]: Found this handy but I could see it be annoying too, if you want to restart the server.

File: db/migrate/20240614183935_add_context_to_users.rb [hunk 3]
```diff
@@ -0,0 +1,5 @@
+class AddContextToUsers < ActiveRecord::Migration[7.1]
+  def change
+    add_column :users, :context, :string
```
Comment (srb-) [inline db/migrate/20240614183935_add_context_to_users.rb:3]: I see that `hangar` uses a separate `account_audits` class for permissions. Didn't think that was necessary to recreate here

**Comment by srb-** [comment]
I'm wondering if this is even needed, as now that I can browse the `hangar` source, it doesn't look like policies are ever applied directly to `noticed` classes? But I could be wrong.

File: app/views/aircrafts/show.html.erb [hunk 4]
```diff
@@ -13,7 +13,11 @@
 
   <div class="d-flex justify-content-between">
     <div>
-      <%= link_to "Edit this aircraft", edit_aircraft_path(@aircraft), class: "btn btn-primary me-2" %>
+
+      <% if policy(@aircraft).create? %>
```
Comment (ddnguyen93) [inline app/views/aircrafts/show.html.erb:17]:
```suggestion
      <% if policy(@aircraft).edit? %>
```

File: app/policies/aircraft_policy.rb [hunk 5]
```diff
@@ -0,0 +1,14 @@
...
+  end
+  
+  def index?
+    true
+  end
+
+  def show?
+    true
+  end
+
```
Comment (ddnguyen93) [inline app/policies/aircraft_policy.rb:13]:
```suggestion
def update?
  create?
end
```

File: app/controllers/aircrafts_controller.rb [hunk 6]
```diff
@@ -8,6 +8,7 @@ def index
 
   # GET /aircrafts/1 or /aircrafts/1.json
   def show
+    authorize(Aircraft)
```
Comment (ddnguyen93) [inline app/controllers/aircrafts_controller.rb:11]:
```suggestion
    authorize(@aircraft)
```
add this to new, edit, update and destroy

**Comment by pJeyakumar** [comment]
@srb- it would be useful to have a `toggle` in the `/users/:id/edit` page to allow us to grant the user the necessary permissions in the policy (e.g. "app:dispatch:wing_admin").
- instead of having to manually update the `User` record in `rails console` to grant the user with the permission (if we were trying to test on localhost)

File: app/models/user.rb [hunk 7]
```diff
@@ -29,6 +29,14 @@ def to_s
     "#{first_name} #{last_name}"
   end
 
+  def admin
+    (context == AircraftPolicy::AIRCRAFT_PERM_STRING) ? "1" : "0"
+  end
+
+  def admin=(value)
+    self.context = (value == "1") ? AircraftPolicy::AIRCRAFT_PERM_STRING : ""
```
Comment (pJeyakumar) [inline app/models/user.rb:37]:
```suggestion
    self.context = AircraftPolicy::WING_ADMIN_PERMISSION if value == "1"
```

**Review by pJeyakumar — CHANGES_REQUESTED** [review]
Some changes to be made to make things a bit nicer.

Hanger tackles permissions quite differently, but I believe it may not be worth replicating the entire `Permissions` structure so that we can move onto the Noticed 2.3.2 upgrade (thoughts @jonathanloos ?)

File: app/models/user.rb [hunk 8]
```diff
@@ -29,6 +29,14 @@ def to_s
     "#{first_name} #{last_name}"
   end
 
+  def admin
+    (context == AircraftPolicy::AIRCRAFT_PERM_STRING) ? "1" : "0"
```
[the remark this answers is not in the document]
Comment (srb-) [reply app/models/user.rb:33]: Yeah, this was being a bit sloppy. The checkbox needs a "1" or "0", but I should put that conversion closer to the view layer. Will do.

File: app/models/user.rb (hunk 7 above)
Comment (srb-) [reply app/models/user.rb:37]: Will this change clear the value, when Admin is taken away?

Comment (pJeyakumar) [reply app/models/user.rb:37]: Ah yes, the case where the user's admin permission is revoked slipped my mind.
- in that case then
```suggestion
    self.context = (value == SET_ADMIN) ? AircraftPolicy::WING_ADMIN_PERMISSION : ""
```
where `SET_ADMIN = 1` is in `user.rb`.

File: app/models/user.rb (hunk 8 above)
[the remark this answers is not in the document]
Comment (srb-) [reply app/models/user.rb]: @pJeyakumar yikes I had forgotten that HTML checkbox behaviour was kinda ugly. OK updated the PR with your suggestions.

**Review by pJeyakumar — APPROVED** [review]

**Review by hvillero — APPROVED** [review]
Looks good to me!!

**Pull request merged**